File: csharp-dotnet/tests/Architecture.Shell.Cqrs.Tests/Integration/CommandExecutionTests.cs
// Copyright (c) 2025 Universal DDD Architecture. All rights reserved.

using System;
using System.Threading;
using System.Threading.Tasks;
using Architecture.Core.Functional;
using Architecture.Shell.Cqrs.Behaviors;
using Architecture.Shell.Cqrs.Tests.TestHelpers;
using Xunit;

namespace Architecture.Shell.Cqrs.Tests.Integration;

/// <summary>
/// Integration tests for command execution lifecycle.
/// Validates transaction management, commit, and rollback behavior.
/// </summary>
public sealed class CommandExecutionTests
{
    /// <summary>
    /// IT-001: Should_CommitTransaction_When_CommandSucceeds.
    /// Validates that successful command execution commits the transaction.
    /// </summary>
    [Fact]
    public async Task Should_CommitTransaction_When_CommandSucceeds()
    {
        // Given: Mediator with UnitOfWork behavior
        var unitOfWork = new InMemoryUnitOfWork();
        var handler = new TestCommandHandler();
        var behavior = new UnitOfWorkBehavior<TestCommand, Result<string>>(unitOfWork);

        var serviceProvider = TestServiceProvider.CreateBuilder()
            .AddHandler<TestCommand, Result<string>>(handler)
            .AddBehavior<TestCommand, Result<string>>(behavior)
            .Build();

        var logger = new TestLogger<IMediator>();
        var mediator = new Mediator(serviceProvider, logger);
        var command = new TestCommand("test data");

        // When: Command is executed successfully
        var result = await mediator.SendAsync<Result<string>>(command, CancellationToken.None);

        // Then: Transaction is committed
        Assert.True(result.IsSuccess);
        Assert.True(unitOfWork.IsCommitted);
        Assert.False(unitOfWork.IsRolledBack);
        Assert.False(unitOfWork.HasActiveTransaction);
    }

    /// <summary>
    /// IT-002: Should_RollbackTransaction_When_CommandThrowsException.
    /// Validates that exceptions trigger transaction rollback.
    /// </summary>
    [Fact]
    public async Task Should_RollbackTransaction_When_CommandThrowsException()
    {
        // Given: Mediator with UnitOfWork behavior and handler that throws
        var unitOfWork = new InMemoryUnitOfWork();
        var handler = new FailingCommandHandler();
        var behavior = new UnitOfWorkBehavior<FailingCommand, Result<string>>(unitOfWork);

        var serviceProvider = TestServiceProvider.CreateBuilder()
            .AddHandler<FailingCommand, Result<string>>(handler)
            .AddBehavior<FailingCommand, Result<string>>(behavior)
            .Build();

        var logger = new TestLogger<IMediator>();
        var mediator = new Mediator(serviceProvider, logger);
        var command = new FailingCommand(ShouldThrow: true);

        // When: Command throws exception
        // Then: Transaction is rolled back
        await Assert.ThrowsAsync<InvalidOperationException>(
            () => mediator.SendAsync<Result<string>>(command, CancellationToken.None));

        Assert.True(unitOfWork.IsRolledBack);
        Assert.False(unitOfWork.IsCommitted);
        Assert.False(unitOfWork.HasActiveTransaction);
    }

    /// <summary>
    /// IT-003: Should_ShareTransaction_When_NestedCommandCalled.
    /// Validates that nested commands reuse the active transaction.
    /// </summary>
    [Fact]
    public async Task Should_ShareTransaction_When_NestedCommandCalled()
    {
        // Given: Mediator with UnitOfWork behavior and nested command handler
        var unitOfWork = new InMemoryUnitOfWork();

        // Create handlers
        var innerHandler = new InnerCommandHandler();

        // Create two separate mediators for proper dependency injection
        var innerServiceProvider = TestServiceProvider.CreateBuilder()
            .AddHandler<InnerCommand, Result<string>>(innerHandler)
            .AddBehavior<InnerCommand, Result<string>>(new UnitOfWorkBehavior<InnerCommand, Result<string>>(unitOfWork))
            .Build();

        var innerLogger = new TestLogger<IMediator>();
        var innerMediator = new Mediator(innerServiceProvider, innerLogger);

        // Outer handler needs the mediator
        var outerHandler = new OuterCommandHandler(innerMediator);

        var outerServiceProvider = TestServiceProvider.CreateBuilder()
            .AddHandler<OuterCommand, Result<string>>(outerHandler)
            .AddBehavior<OuterCommand, Result<string>>(new UnitOfWorkBehavior<OuterCommand, Result<string>>(unitOfWork))
            .Build();

        var outerLogger = new TestLogger<IMediator>();
        var outerMediator = new Mediator(outerServiceProvider, outerLogger);

        var command = new OuterCommand("outer", "inner");

        // When: Outer command executes and sends inner command
        var result = await outerMediator.SendAsync<Result<string>>(command, CancellationToken.None);

        // Then: Transaction is committed once (shared across both commands)
        Assert.True(result.IsSuccess);
        Assert.True(unitOfWork.IsCommitted);
        Assert.False(unitOfWork.HasActiveTransaction);
        Assert.Contains("Outer:", result.Value);
        Assert.Contains("Inner:", result.Value);
    }

    /// <summary>
    /// IT-008: Should_CommitTransaction_When_HandlerReturnsResultFailure.
    /// Validates BR-007/BR-008: Business failures (Result.Failure) commit transaction.
    /// </summary>
    [Fact]
    public async Task Should_CommitTransaction_When_HandlerReturnsResultFailure()
    {
        // Given: Mediator with UnitOfWork behavior and handler that returns Result.Failure
        var unitOfWork = new InMemoryUnitOfWork();
        var handler = new FailingCommandHandler();
        var behavior = new UnitOfWorkBehavior<FailingCommand, Result<string>>(unitOfWork);

        var serviceProvider = TestServiceProvider.CreateBuilder()
            .AddHandler<FailingCommand, Result<string>>(handler)
            .AddBehavior<FailingCommand, Result<string>>(behavior)
            .Build();

        var logger = new TestLogger<IMediator>();
        var mediator = new Mediator(serviceProvider, logger);
        var command = new FailingCommand(ShouldThrow: false);

        // When: Handler returns Result.Failure (business error)
        var result = await mediator.SendAsync<Result<string>>(command, CancellationToken.None);

        // Then: Transaction is committed (business failure is valid state)
        Assert.True(result.IsFailure);
        Assert.True(unitOfWork.IsCommitted);
        Assert.False(unitOfWork.IsRolledBack);
        Assert.False(unitOfWork.HasActiveTransaction);
    }

    /// <summary>
    /// IT-008b: Should_CommitTransaction_When_VoidCommandHandlerReturnsResultFailure.
    /// Validates BR-007/BR-008 for void commands returning Result.Failure.
    /// </summary>
    [Fact]
    public async Task Should_CommitTransaction_When_VoidCommandHandlerReturnsResultFailure()
    {
        // Given: Mediator with void command handler that returns failure
        var unitOfWork = new InMemoryUnitOfWork();
        var handler = new VoidCommandHandlerWithFailure();
        var behavior = new UnitOfWorkBehavior<VoidTestCommand, Result>(unitOfWork);

        var serviceProvider = TestServiceProvider.CreateBuilder()
            .AddHandler<VoidTestCommand, Result>(handler)
            .AddBehavior<VoidTestCommand, Result>(behavior)
            .Build();

        var logger = new TestLogger<IMediator>();
        var mediator = new Mediator(serviceProvider, logger);
        var command = new VoidTestCommand("test");

        // When: Handler returns Result.Failure
        var result = await mediator.SendAsync(command, CancellationToken.None);

        // Then: Transaction is committed
        Assert.True(result.IsFailure);
        Assert.True(unitOfWork.IsCommitted);
        Assert.False(unitOfWork.IsRolledBack);
    }

    /// <summary>
    /// IT-010: Should_RollbackTransaction_When_BehaviorThrowsException.
    /// Validates spec.md:L68-69 edge case - behavior exceptions trigger rollback.
    /// </summary>
    [Fact]
    public async Task Should_RollbackTransaction_When_BehaviorThrowsException()
    {
        // Given: Mediator with behavior that throws
        var unitOfWork = new InMemoryUnitOfWork();
        var handler = new TestCommandHandler();
        var unitOfWorkBehavior = new UnitOfWorkBehavior<TestCommand, Result<string>>(unitOfWork, order: 10);
        var throwingBehavior = new ThrowingBehavior<TestCommand, Result<string>>(order: 20);

        var serviceProvider = TestServiceProvider.CreateBuilder()
            .AddHandler<TestCommand, Result<string>>(handler)
            .AddBehavior<TestCommand, Result<string>>(unitOfWorkBehavior)
            .AddBehavior<TestCommand, Result<string>>(throwingBehavior)
            .Build();

        var logger = new TestLogger<IMediator>();
        var mediator = new Mediator(serviceProvider, logger);
        var command = new TestCommand("test");

        // When: Behavior throws exception
        // Then: Transaction is rolled back
        await Assert.ThrowsAsync<InvalidOperationException>(
            () => mediator.SendAsync<Result<string>>(command, CancellationToken.None));

        Assert.True(unitOfWork.IsRolledBack);
        Assert.False(unitOfWork.IsCommitted);
    }

    /// <summary>
    /// Helper handler that returns Result.Fail for void commands.
    /// </summary>
    private sealed class VoidCommandHandlerWithFailure : ICommandHandler<VoidTestCommand>
    {
        public Task<Result> HandleAsync(VoidTestCommand request, CancellationToken cancellationToken)
        {
            return Task.FromResult(Result.Fail(Error.Domain("VOID_ERROR", "Void command business failure")));
        }
    }

    /// <summary>
    /// Helper behavior that throws exception.
    /// </summary>
    private sealed class ThrowingBehavior<TRequest, TResponse> : IPipelineBehavior<TRequest, TResponse>
        where TRequest : IBaseRequest
    {
        public ThrowingBehavior(int order = 50)
        {
            Order = order;
        }

        public int Order { get; }

        public Task<TResponse> HandleAsync(TRequest request, RequestHandlerDelegate<TResponse> continuation, CancellationToken cancellationToken)
        {
            throw new InvalidOperationException("Simulated behavior exception");
        }
    }
}

File: java-spring/architecture-shell-cqrs/src/test/java/com/architecture/shell/cqrs/CommandExecutionTests.java
package com.architecture.shell.cqrs;

import com.architecture.core.functional.Result;
import org.junit.jupiter.api.Test;

import java.util.UUID;
import java.util.concurrent.atomic.AtomicInteger;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;

/**
 * Integration tests for Command Execution Lifecycle.
 * Tests validate end-to-end behavior including transaction management.
 * Naming: Should_ExpectedBehavior_When_StateUnderTest
 */
class CommandExecutionTests {

    /**
     * T148: IT-001 - Command Execution Lifecycle (Success)
     * Given: Command handler that succeeds
     * When: Command is executed through mediator with UnitOfWork behavior
     * Then: Should commit transaction on successful completion
     */
    @Test
    void should_CommitTransaction_When_CommandSucceeds() {
        // Given: InMemory UnitOfWork and successful command handler
        var unitOfWork = new InMemoryUnitOfWork();
        var handlerRegistry = new HandlerRegistry();
        handlerRegistry.register(new SuccessCommandHandler());

        var behaviorRegistry = new BehaviorRegistry();
        behaviorRegistry.register(new com.architecture.shell.cqrs.behaviors.UnitOfWorkBehavior<>(unitOfWork), 30);

        var mediator = new MediatorImpl(handlerRegistry, behaviorRegistry);

        // When: Command is executed
        var result = mediator.<Result<String>>send(new SuccessCommand());

        // Then: Transaction should be committed
        assertThat(result.isSuccess()).isTrue();
        assertThat(result.getValue()).isEqualTo("Success");
        assertThat(unitOfWork.wasCommitted()).isTrue();
        assertThat(unitOfWork.wasRolledBack()).isFalse();
    }

    /**
     * T152: IT-002 - Command Execution Rollback (Exception)
     * Given: Command handler that throws exception
     * When: Command is executed
     * Then: Should rollback transaction when handler throws exception
     */
    @Test
    void should_RollbackTransaction_When_CommandThrowsException() {
        // Given: InMemory UnitOfWork and failing command handler
        var unitOfWork = new InMemoryUnitOfWork();
        var handlerRegistry = new HandlerRegistry();
        handlerRegistry.register(new ThrowingCommandHandler());

        var behaviorRegistry = new BehaviorRegistry();
        behaviorRegistry.register(new com.architecture.shell.cqrs.behaviors.UnitOfWorkBehavior<>(unitOfWork), 30);

        var mediator = new MediatorImpl(handlerRegistry, behaviorRegistry);

        // When/Then: Command execution should throw exception
        assertThatThrownBy(() -> mediator.send(new ThrowingCommand()))
            .isInstanceOf(RuntimeException.class)
            .hasMessageContaining("Infrastructure error");

        // Then: Transaction should be rolled back
        assertThat(unitOfWork.wasCommitted()).isFalse();
        assertThat(unitOfWork.wasRolledBack()).isTrue();
    }

    /**
     * T157: IT-003 - Nested Command Transaction Reuse
     * Given: Outer command with active transaction
     * When: Nested command is executed within same transaction scope
     * Then: Should share transaction between outer and nested commands
     */
    @Test
    void should_ShareTransaction_When_NestedCommandCalled() {
        // Given: InMemory UnitOfWork and nested command scenario
        var unitOfWork = new InMemoryUnitOfWork();
        var handlerRegistry = new HandlerRegistry();

        var mediatorRef = new Object() {
            MediatorImpl value;
        };

        handlerRegistry.register(new CommandHandler<OuterCommand, Result<Void>>() {
            @Override
            public Result<Void> handle(OuterCommand command) {
                // Send nested command within transaction
                mediatorRef.value.send(new NestedCommand());
                return Result.success();
            }
        });

        handlerRegistry.register(new NestedCommandHandler());

        var behaviorRegistry = new BehaviorRegistry();
        behaviorRegistry.register(new com.architecture.shell.cqrs.behaviors.UnitOfWorkBehavior<>(unitOfWork), 30);

        mediatorRef.value = new MediatorImpl(handlerRegistry, behaviorRegistry);

        // When: Outer command is executed
        mediatorRef.value.send(new OuterCommand());

        // Then: Transaction should be shared (single commit, no nested transactions)
        assertThat(unitOfWork.getBeginTransactionCount()).isEqualTo(1); // Only one transaction opened
        assertThat(unitOfWork.wasCommitted()).isTrue();
        assertThat(unitOfWork.wasRolledBack()).isFalse();
    }

    /**
     * T211: IT-008 - Transaction Commit on Business Failure
     * Given: Command handler that returns Result.Failure (business error)
     * When: Handler returns failure result
     * Then: Should commit transaction (business failures are valid states per BR-008)
     */
    @Test
    void should_CommitTransaction_When_HandlerReturnsResultFailure() {
        // Given: InMemory UnitOfWork and handler that returns business failure
        var unitOfWork = new InMemoryUnitOfWork();
        var handlerRegistry = new HandlerRegistry();
        handlerRegistry.register(new BusinessFailureCommandHandler());

        var behaviorRegistry = new BehaviorRegistry();
        behaviorRegistry.register(new com.architecture.shell.cqrs.behaviors.UnitOfWorkBehavior<>(unitOfWork), 30);

        var mediator = new MediatorImpl(handlerRegistry, behaviorRegistry);

        // When: Command is executed
        var result = mediator.<Result<String>>send(new BusinessFailureCommand());

        // Then: Result is failure but transaction is COMMITTED (not rolled back)
        assertThat(result.isFailure()).isTrue();
        assertThat(result.getError().getMessage()).contains("Business validation failed");
        assertThat(unitOfWork.wasCommitted()).isTrue();
        assertThat(unitOfWork.wasRolledBack()).isFalse();
    }

    /**
     * T211b: IT-008b - Transaction Commit on Business Failure (Void Commands)
     * Given: Void command handler that returns Result<Unit>.Failure
     * When: Handler returns failure result
     * Then: Should commit transaction (extends IT-008 for void commands)
     */
    @Test
    void should_CommitTransaction_When_VoidCommandHandlerReturnsResultFailure() {
        // Given: InMemory UnitOfWork and void command handler that returns business failure
        var unitOfWork = new InMemoryUnitOfWork();
        var handlerRegistry = new HandlerRegistry();
        handlerRegistry.register(new VoidBusinessFailureCommandHandler());

        var behaviorRegistry = new BehaviorRegistry();
        behaviorRegistry.register(new com.architecture.shell.cqrs.behaviors.UnitOfWorkBehavior<>(unitOfWork), 30);

        var mediator = new MediatorImpl(handlerRegistry, behaviorRegistry);

        // When: Void command is executed
        var result = mediator.<Result<Void>>send(new VoidBusinessFailureCommand());

        // Then: Result is failure but transaction is COMMITTED
        assertThat(result.isFailure()).isTrue();
        assertThat(result.getError().getMessage()).contains("Void command business validation failed");
        assertThat(unitOfWork.wasCommitted()).isTrue();
        assertThat(unitOfWork.wasRolledBack()).isFalse();
    }

    /**
     * T222: IT-010 - Behavior Exception Rollback
     * Given: Pipeline behavior that throws exception
     * When: Behavior throws during execution
     * Then: Should rollback transaction (extends IT-002 beyond handler exceptions)
     */
    @Test
    void should_RollbackTransaction_When_BehaviorThrowsException() {
        // Given: InMemory UnitOfWork and behavior that throws exception
        var unitOfWork = new InMemoryUnitOfWork();
        var handlerRegistry = new HandlerRegistry();
        handlerRegistry.register(new SuccessCommandHandler());

        var behaviorRegistry = new BehaviorRegistry();
        behaviorRegistry.register(new com.architecture.shell.cqrs.behaviors.UnitOfWorkBehavior<>(unitOfWork), 20); // Transaction first
        behaviorRegistry.register(new ThrowingBehavior(), 30); // Throwing behavior after transaction

        var mediator = new MediatorImpl(handlerRegistry, behaviorRegistry);

        // When/Then: Behavior exception should propagate
        assertThatThrownBy(() -> mediator.send(new SuccessCommand()))
            .isInstanceOf(RuntimeException.class)
            .hasMessageContaining("Behavior threw exception");

        // Then: Transaction should be rolled back
        assertThat(unitOfWork.wasCommitted()).isFalse();
        assertThat(unitOfWork.wasRolledBack()).isTrue();
    }

    // Test fixtures

    private static class SuccessCommand implements Command<Result<String>> {
    }

    private static class SuccessCommandHandler implements CommandHandler<SuccessCommand, Result<String>> {
        @Override
        public Result<String> handle(SuccessCommand command) {
            return Result.success("Success");
        }
    }

    private static class ThrowingCommand implements Command<Result<Void>> {
    }

    private static class ThrowingCommandHandler implements CommandHandler<ThrowingCommand, Result<Void>> {
        @Override
        public Result<Void> handle(ThrowingCommand command) {
            throw new RuntimeException("Infrastructure error - database unavailable");
        }
    }

    private static class OuterCommand implements Command<Result<Void>> {
    }

    private static class NestedCommand implements Command<Result<Void>> {
    }

    private static class NestedCommandHandler implements CommandHandler<NestedCommand, Result<Void>> {
        @Override
        public Result<Void> handle(NestedCommand command) {
            return Result.success();
        }
    }

    private static class BusinessFailureCommand implements Command<Result<String>> {
    }

    private static class BusinessFailureCommandHandler implements CommandHandler<BusinessFailureCommand, Result<String>> {
        @Override
        public Result<String> handle(BusinessFailureCommand command) {
            // Business validation failure - transaction should commit
            return Result.failure(com.architecture.core.functional.Error.validation(
                "INSUFFICIENT_STOCK",
                "Business validation failed: insufficient stock",
                new java.util.HashMap<>()
            ));
        }
    }

    private static class VoidBusinessFailureCommand implements Command<Result<Void>> {
    }

    private static class VoidBusinessFailureCommandHandler implements CommandHandler<VoidBusinessFailureCommand, Result<Void>> {
        @Override
        public Result<Void> handle(VoidBusinessFailureCommand command) {
            // Void command business validation failure - transaction should commit
            return Result.failure(com.architecture.core.functional.Error.validation(
                "VOID_VALIDATION_FAILURE",
                "Void command business validation failed",
                new java.util.HashMap<>()
            ));
        }
    }

    /**
     * Behavior that throws exception (for IT-010 test)
     */
    private static class ThrowingBehavior implements PipelineBehavior<BaseRequest, Object> {
        @Override
        public Object handle(BaseRequest request, RequestHandlerDelegate<Object> next) {
            throw new RuntimeException("Behavior threw exception");
        }
    }
}
